### PR TITLE
Add Continuous Auto-Scrolling

### PR DESCRIPTION
**Pull request by regulus79**
# Status
This PR is ready to merge.

## Description

This pull request adds the option to have auto-scrolling continuously scroll the view to keep up with the playback.

## Changes
- New icons for Stepped/Continuous auto-scrolling
- `AutoScrollState` enum changed from `Enabled`/`Disabled` to `Stepped`/`Continuous`/`Disabled`
- The auto-scroll button in `TimeLineWidget.cpp` is changed from having 2 states (`Enabled`/`Disabled`) to having 3 states (`Stepped`/`Continuous`/`Disabled`)
- In the code for auto-scrolling in both `SongEditor.cpp` and `PianoRoll.cpp`, a check is added for whether the mode is `Stepped` or `Continuous`; if it's `Stepped`, then is runs the old code for auto-scrolling; if it's `Continuous`, then it snaps the horizontal scroll to halfway behind the time line so that the line appears in the center of the view.
- `SongEditor.cpp`'s horizontal scrollbar's functionality was changed to make it scroll by ticks instead of bars. This required modifying `TrackContentWidget.cpp` to be positioned by ticks instead of bars to properly draw the grid background.
- Additionally, due to the change in scrollbar step size, the scrolling sensitivity for the Song Editor had to be changed to a reasonable rate. In this PR, the scrolling rate behaves as normal and does not depend on the zooming. I had mentioned the possibility of adding that in the discord, but I decided to save it for another PR.

### CODE REVIEW

File: src/gui/editors/PianoRoll.cpp [hunk 1]
```diff
@@ -4463,27 +4463,33 @@ bool PianoRoll::deleteSelectedNotes()
 void PianoRoll::autoScroll( const TimePos & t )
 {
 	const int w = width() - m_whiteKeyWidth;
-	if( t > m_currentPosition + w * TimePos::ticksPerBar() / m_ppb )
+	if (m_timeLine->autoScroll() == TimeLineWidget::AutoScrollState::Stepped) 
 	{
-		m_leftRightScroll->setValue( t.getBar() * TimePos::ticksPerBar() );
+		if( t > m_currentPosition + w * TimePos::ticksPerBar() / m_ppb )
```
Comment (Rossmaxx) [inline src/gui/editors/PianoRoll.cpp:4468]: Style

Comment (Rossmaxx) [reply src/gui/editors/PianoRoll.cpp:4468]: In the 3 following lines too

File: src/gui/editors/SongEditor.cpp [hunk 2]
```diff
@@ -234,7 +234,7 @@ SongEditor::SongEditor( Song * song ) :
 	m_leftRightScroll->setMinimum( 0 );
 	m_leftRightScroll->setMaximum( 0 );
 	m_leftRightScroll->setSingleStep( 1 );
-	m_leftRightScroll->setPageStep( 20 );
+	m_leftRightScroll->setPageStep( 20 * TimePos::ticksPerBar() );
```
Comment (Rossmaxx) [inline src/gui/editors/SongEditor.cpp:237]: Style here too

File: src/gui/editors/SongEditor.cpp [hunk 3]
```diff
@@ -713,7 +713,7 @@ void SongEditor::hideMasterPitchFloat( void )
 
 void SongEditor::updateScrollBar( int len )
 {
-	m_leftRightScroll->setMaximum( len );
+	m_leftRightScroll->setMaximum( len * TimePos::ticksPerBar() );
```
Comment (Rossmaxx) [inline src/gui/editors/SongEditor.cpp:716]: Here too. 
For eg 
```suggestion
	m_leftRightScroll->setMaximum(len * TimePos::ticksPerBar());
```

File: src/gui/editors/SongEditor.cpp [hunk 4]
```diff
@@ -756,22 +756,29 @@ void SongEditor::updatePosition( const TimePos & t )
 	const auto widgetWidth = compactTrackButtons ? DEFAULT_SETTINGS_WIDGET_WIDTH_COMPACT : DEFAULT_SETTINGS_WIDGET_WIDTH;
 	const auto trackOpWidth = compactTrackButtons ? TRACK_OP_WIDTH_COMPACT : TRACK_OP_WIDTH;
 
-	if( ( m_song->isPlaying() && m_song->m_playMode == Song::PlayMode::Song
-		  && m_timeLine->autoScroll() == TimeLineWidget::AutoScrollState::Enabled) ||
+	if( ( m_song->isPlaying() && m_song->m_playMode == Song::PlayMode::Song ) ||
```
Comment (Rossmaxx) [inline src/gui/editors/SongEditor.cpp:759]: There's more. See the rest from the diff and correct it.

**Review by Rossmaxx — COMMENTED** [review]
Minor style review, also pay attention to codefactor. 

Btw you forgot classic theme.

File: src/gui/editors/PianoRoll.cpp [hunk 5]
```diff
@@ -4463,30 +4463,36 @@ bool PianoRoll::deleteSelectedNotes()
 void PianoRoll::autoScroll( const TimePos & t )
 {
 	const int w = width() - m_whiteKeyWidth;
-	if( t > m_currentPosition + w * TimePos::ticksPerBar() / m_ppb )
+	if (m_timeLine->autoScroll() == TimeLineWidget::AutoScrollState::Stepped) 
 	{
-		m_leftRightScroll->setValue( t.getBar() * TimePos::ticksPerBar() );
+		if(t > m_currentPosition + w * TimePos::ticksPerBar() / m_ppb)
```
Comment (Rossmaxx) [inline src/gui/editors/PianoRoll.cpp:4468]: Sorry to bug you on style again, but we use spaces between if and the parantheses. I've seen similar style getting repeated. Correct those too.

**Comment by regulus79** [comment]
Okay, I made some icons for classic, and hopefully I've fixed all the style issues.

**Review by Rossmaxx — APPROVED** [review]
Looks good.

File: src/gui/editors/SongEditor.cpp [hunk 6]
```diff
@@ -542,15 +542,15 @@ void SongEditor::wheelEvent( QWheelEvent * we )
 	}
 
 	// FIXME: Reconsider if determining orientation is necessary in Qt6.
-	else if(abs(we->angleDelta().x()) > abs(we->angleDelta().y())) // scrolling is horizontal
+	else if (abs(we->angleDelta().x()) > abs(we->angleDelta().y())) // scrolling is horizontal
```
Comment (sakertooth) [inline src/gui/editors/SongEditor.cpp:545]:
```suggestion
	else if (std::abs(we->angleDelta().x()) > std::abs(we->angleDelta().y())) // scrolling is horizontal
```

File: src/gui/editors/PianoRoll.cpp [hunk 7]
```diff
@@ -4463,30 +4463,36 @@ bool PianoRoll::deleteSelectedNotes()
...
 	{
-		m_leftRightScroll->setValue( t.getBar() * TimePos::ticksPerBar() );
+		if (t > m_currentPosition + w * TimePos::ticksPerBar() / m_ppb)
+		{
+			m_leftRightScroll->setValue(t.getBar() * TimePos::ticksPerBar());
+		}
+		else if (t < m_currentPosition)
+		{
+			TimePos t2 = qMax(t - w * TimePos::ticksPerBar() *
+						TimePos::ticksPerBar() / m_ppb, (tick_t) 0);
```
Comment (sakertooth) [inline src/gui/editors/PianoRoll.cpp:4475]:
```suggestion
			TimePos t2 = std::max(t - w * TimePos::ticksPerBar() *
						TimePos::ticksPerBar() / m_ppb, static_cast<tick_t>(0));
```

File: src/gui/editors/SongEditor.cpp [hunk 8]
```diff
@@ -756,22 +756,29 @@ void SongEditor::updatePosition( const TimePos & t )
...
-			animateScroll( m_leftRightScroll, t.getBar(), m_smoothScroll );
+			if (t > m_currentPosition + w * TimePos::ticksPerBar()
+								/ pixelsPerBar())
+			{
+				animateScroll(m_leftRightScroll, t.getTicks(), m_smoothScroll);
+			}
+			else if (t < m_currentPosition)
+			{
+				animateScroll(m_leftRightScroll, t.getTicks(), m_smoothScroll);
+			}
```
Comment (sakertooth) [inline src/gui/editors/SongEditor.cpp:777]: I think this is simpler and should work the same way. Feel free to change the name of `nextPosition` if that's not accurate.
```suggestion
			const auto nextPosition = m_currentPosition + w * TimePos::ticksPerBar() / pixelsPerBar();
			if (t <= nextPosition && t >= m_currentPosition) { return; }
			animateScroll(m_leftRightScroll, t.getTicks(), m_smoothScroll);
```

**Review by sakertooth — COMMENTED** [review]
We should do testing before merge.

File: src/gui/editors/PianoRoll.cpp [hunk 9]
```diff
@@ -4062,7 +4062,7 @@ void PianoRoll::stop()
 {
 	Engine::getSong()->stop();
 	m_recording = false;
-	m_scrollBack = ( m_timeLine->autoScroll() == TimeLineWidget::AutoScrollState::Enabled );
+	m_scrollBack = (m_timeLine->autoScroll() != TimeLineWidget::AutoScrollState::Disabled);
```
Comment (michaelgregorius) [inline src/gui/editors/PianoRoll.cpp:4065]: Parentheses are not needed here.

**Comment by michaelgregorius** [comment]
I have found a bug that's introduced by this PR, i.e. that is not in master.

## Steps to reproduce
1. Load a song.
2. Set the mode to "Continuous".
3. Play the song.
4. Quickly change to "Off" and then to "Paged".

The line of the play head line will not move with the play head anymore.

**Review by michaelgregorius — CHANGES_REQUESTED** [review]
The bug that's described in my previous comment should be fixed.

**Comment by regulus79** [comment]
I think I found the problem. The return statement in the simplification of the stepped auto scrolling code exited the function too soon, which happened to skip the code for updating the play head line position. I fixed it by modifying the if statement to not require a return.

**Review by michaelgregorius — APPROVED** [review]

**Comment by sakertooth** [comment]
> I think I found the problem. The return statement in the simplification of the stepped auto scrolling code exited the function too soon, which happened to skip the code for updating the play head line position. I fixed it by modifying the if statement to not require a return.

Had a feeling this would break something, my bad. I got tunnel vision on that one piece of code and forgot about the surrounding function.

**Comment by sakertooth** [comment]
I think it might be helpful to have more descriptions for the tooltip when toggling between the modes (i.e., you can tell the user what this mode will do in particular rather than simply saying "Auto scrolling", because not all modes auto scroll).

File: src/gui/editors/SongEditor.cpp [hunk 10]
```diff
@@ -756,22 +756,25 @@ void SongEditor::updatePosition( const TimePos & t )
 	const auto widgetWidth = compactTrackButtons ? DEFAULT_SETTINGS_WIDGET_WIDTH_COMPACT : DEFAULT_SETTINGS_WIDGET_WIDTH;
 	const auto trackOpWidth = compactTrackButtons ? TRACK_OP_WIDTH_COMPACT : TRACK_OP_WIDTH;
 
-	if( ( m_song->isPlaying() && m_song->m_playMode == Song::PlayMode::Song
-		  && m_timeLine->autoScroll() == TimeLineWidget::AutoScrollState::Enabled) ||
-							m_scrollBack == true )
+	if ((m_song->isPlaying() && m_song->m_playMode == Song::PlayMode::Song)
+							|| m_scrollBack == true)
```
Comment (sakertooth) [inline src/gui/editors/SongEditor.cpp:760]:
```suggestion
							|| m_scrollBack)
```

**Review by sakertooth — APPROVED** [review]
LGTM, also something minor I noticed

File: src/gui/editors/TimeLineWidget.cpp [hunk 11]
```diff
@@ -88,9 +88,10 @@ void TimeLineWidget::setXOffset(const int x)
 void TimeLineWidget::addToolButtons( QToolBar * _tool_bar )
 {
 	auto autoScroll = new NStateButton(_tool_bar);
-	autoScroll->setGeneralToolTip( tr( "Auto scrolling" ) );
-	autoScroll->addState( embed::getIconPixmap( "autoscroll_on" ) );
-	autoScroll->addState( embed::getIconPixmap( "autoscroll_off" ) );
+	autoScroll->setGeneralToolTip(tr("Auto scrolling"));
+	autoScroll->addState(embed::getIconPixmap("autoscroll_stepped_on"), "Stepped auto scrolling");
+	autoScroll->addState(embed::getIconPixmap("autoscroll_continuous_on"), "Continuous auto scrolling");
+	autoScroll->addState(embed::getIconPixmap("autoscroll_off"), "Auto scrolling disabled");
```
Comment (DomClark) [inline src/gui/editors/TimeLineWidget.cpp:94]: These tooltips should be translatable:
```suggestion
	autoScroll->addState(embed::getIconPixmap("autoscroll_stepped_on"), tr("Stepped auto scrolling"));
	autoScroll->addState(embed::getIconPixmap("autoscroll_continuous_on"), tr("Continuous auto scrolling"));
	autoScroll->addState(embed::getIconPixmap("autoscroll_off"), tr("Auto scrolling disabled"));
```

**Comment by regulus79** [comment]
I realized that changing the scrolling step size also broke zooming centered around the mouse position. I have now fixed that; just a simple multiplication by `TimePos::ticksPerBar()`.

**Comment by regulus79** [comment]
Just remembered that there was another bug when drawing the rubberband selection rect, where the left side would get teleported to the left as everything scrolled. Once again, a simple division by `TimePos::ticksPerBar()` fixed it.

**Comment by regulus79** [comment]
Alright, I think those are all the necessary changes, so unless anyone spots a new issue, this should pr should be ready to merge.

**Review by michaelgregorius — APPROVED** [review]

**Comment by Gabrielxd195** [comment]
I have a question about this, and it is that the continuous scrolling is more fluid and consistent in the Piano Roll than in the Song Editor, since in the latter the scrolling stops suddenly and is not constant. I show you this in the following videos:

Autoscrolling in Piano Roll

https://github.com/user-attachments/assets/af263d05-8b0a-41ac-990e-766862c6f637

Autoscrolling in Song Editor

https://github.com/user-attachments/assets/5af31285-9d37-4554-8600-2e24f58a5b27

Is this on purpose? Or is there some other way to improve this?

**Comment by regulus79** [comment]
Ah, whoops, I forgot to test it with smooth scroll enabled. I should have set the scroll using `m_leftRightScroll->setValue()` instead of `animateScroll()`. 

For now, you can temporarily fix it by disabling smooth scroll in the settings.

**Pull request merged**